File: snippets/design-patterns/structural/decorator.go
// --- Decorator Pattern ---
// Use Case: Dynamically add behavior to objects without modifying their code.
// Example: Middleware in web frameworks, logging, or caching.
// Description: The Decorator Pattern attaches additional responsibilities to an object dynamically,
// allowing behavior to be added or modified at runtime.

package main

import "fmt"

type Component interface {
	Operation() string
}

type ConcreteComponent struct{}

func (c *ConcreteComponent) Operation() string {
	return "ConcreteComponent"
}

type Decorator struct {
	component Component
}

func (d *Decorator) Operation() string {
	return fmt.Sprintf("%s + Decorator", d.component.Operation())
}

func main() {
	component := &ConcreteComponent{}
	decorator := &Decorator{component: component}

	fmt.Println(decorator.Operation()) // Output: ConcreteComponent + Decorator
}

File: snippets/design-patterns/structural/decorator.py
'''
--- Decorator Pattern ---
Use Case: When you need to add functionality to objects dynamically.
Example: Logging, authentication, or caching.
'''

def decorator(func):
    def wrapper(*args, **kwargs):
        print(f"Function {func.__name__} is being called")
        result = func(*args, **kwargs)
        print(f"Function {func.__name__} returned {result}")
        return result
    return wrapper

@decorator
def add(a, b):
    return a + b

# Example Usage
print(add(5, 3))
# Output:
# Function add is being called
# Function add returned 8
# 8